### ISSUE TITLE
performance improvement for list request with labelSelector

### ISSUE THREAD

**Issue by likakuli** [karmada-io/karmada#2234]
**What would you like to be added**:

Use index to improve the performance of list request with labelSelector

**Why is this needed**:

LabelSelector is used in many places in `karmada-controller` to list object. `controller-runtime` not setup index for label and not support to setup index for label too. So the list request cost O(n). When there are many object, it will cost too much time.

controller-runtime code: https://github.com/kubernetes-sigs/controller-runtime/blob/f46919744bee01060c9084a285e049afffd38c9d/pkg/cache/internal/cache_reader.go#L115-L138

**Comment by likakuli** [comment]
I don't know whether this function should be implemented in `controller-runtime` or just in `karmada-controller`.  
There are two points related to this.

1. `controller-runtime` not support index when to use labelSelector in list request.
2. `controller-runtime` not support complicated fieldSelector just as the comment. 

In our case, we implemented this by two steps.

1. We add custom manager and cacher used to convert labelSelector to fieldSelector.
2. We use custom `client-go` and `controller-runtime` package to support complicated selector. refer to:  
https://github.com/kubernetes/kubernetes/pull/109334 
https://github.com/kubernetes-sigs/controller-runtime/pull/1838

**Comment by RainbowMango** [comment]
cc @yy158775 Please take a look this issue.

**Comment by yy158775** [comment]
From my point of view,I think we can convet labelSelector to fieldSelector with a custom IndexFunc.The IndexFunc can compute the set of indexed values for an object's labels.

Eg: If we want to select resourcebindings with label key "clusterpropagationpolicy.karmada.io/name" and specific value,we can construct a IndexFunc which return "key=value" pair when the "clusterpropagationpolicy.karmada.io/name" exists in object labels.

Such as:
indices["labels"] = {"clusterpropagationpolicy.karmada.io/name=$var":[key1,key2,key3,......],......}
indexers["labels"] = indexfunc

If we need to select objects by other labels,we can edit the indexfunc to support.

**Comment by likakuli** [comment]
If no one is working on it, i can open a pr to support this issue.

**Comment by RainbowMango** [comment]
Sure thing! (anyone is welcome to pick a task without an assignment)

Have you done it on your side?

I know @yy158775 started looking at it and working on the demo.

This issue is much more involved. And @yy158775 can cooperate with you.

**Comment by likakuli** [comment]
Our internal version has implemented this function.

**Comment by yy158775** [comment]
I think this is worth optimizing.So I look foward to seeing your version.

**Comment by RainbowMango** [comment]
/assign @likakuli 
So, please help to make a PR for this.

**Comment by yy158775** [comment]
> Our internal version has implemented this function.

Are you convenient to open a PR? I am ready to work on this issue with you.

**Comment by likakuli** [comment]
sorry, i will open a pr this weekend. or you can open a pr if you have already completed it. thx

**Comment by weilaaa** [comment]
In my opinion, we need to fundamentally enhance the capabilities of client-go. See proposal issues: https://github.com/kubernetes/kubernetes/issues/109329

**Comment by RainbowMango** [comment]
@weilaaa Are you suggesting waiting for the proposed issue in K/K?

**Comment by weilaaa** [comment]
Yes, it may be the most appropriate way, but I can't guarantee that this proposal will be passed. At least we can wait for the conclusion of the community(K/K) on this proposal.

**Comment by zach593** [comment]
At the moment, it seems that there are only a few remaining places where List operations using labelSelector could be replaced with index. I’ve been quite busy recently, so I haven’t been keeping up with the community updates. I’ll take a closer look soon and work on optimizing these remaining cases one by one.
